Mosek: Add eps for optimality tolerance parameters

Paulnkk opened the pull request:
## Description
I added a parameter 'eps' which automatically maps a user-defined tolerance to Mosek optimality tolerance parameters. 
Issue link (if applicable):
I found the issue here: https://github.com/cvxpy/cvxpy/issues/1206

## Type of change
- [ ] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [x] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [ ] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cvxpy/cvxpy?pullRequest=2119) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cvxpy/cvxpy?pullRequest=2119) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cvxpy/cvxpy?pullRequest=2119) it.</sub>

Paulnkk commented:
There are two things I am curious about:

1. For now, I hardcoded the optimality tolerance parameters from https://docs.mosek.com/7.0/toolbox/Parameters.html but just for conic solvers. Should I also add the other optimality tolerance parameters (some might be irrelevant depending on the solver used).
2. Potentially, it may be also useful to implement a test for the 'eps' parameter.

rileyjmurray commented:
I'd like for us to reference a more recent version of the MOSEK docs, as well as the Python-specific docs:

https://docs.mosek.com/9.3/pythonapi/param-groups.html

Conic interior-point method

[dparam.intpnt_co_tol_dfeas](https://docs.mosek.com/9.3/pythonapi/parameters.html#mosek.dparam.intpnt_co_tol_dfeas)
[dparam.intpnt_co_tol_infeas](https://docs.mosek.com/9.3/pythonapi/parameters.html#mosek.dparam.intpnt_co_tol_infeas)
[dparam.intpnt_co_tol_mu_red](https://docs.mosek.com/9.3/pythonapi/parameters.html#mosek.dparam.intpnt_co_tol_mu_red)
[dparam.intpnt_co_tol_pfeas](https://docs.mosek.com/9.3/pythonapi/parameters.html#mosek.dparam.intpnt_co_tol_pfeas)
[dparam.intpnt_co_tol_rel_gap](https://docs.mosek.com/9.3/pythonapi/parameters.html#mosek.dparam.intpnt_co_tol_rel_gap)

Simplex optimizer

[dparam.basis_rel_tol_s](https://docs.mosek.com/9.3/pythonapi/parameters.html#mosek.dparam.basis_rel_tol_s)
[dparam.basis_tol_s](https://docs.mosek.com/9.3/pythonapi/parameters.html#mosek.dparam.basis_tol_s)
[dparam.basis_tol_x](https://docs.mosek.com/9.3/pythonapi/parameters.html#mosek.dparam.basis_tol_x)

I think we should set all of these parameters when the user provides ``eps``. @aszekMosek, do you agree? If this seems like too blunt of a solution, perhaps the parameter name could be ``eps_all``, to indicate how broadly the setting applies.

Paulnkk commented:
@aszekMosek thanks very much for the suggestion!

Paulnkk commented:
I added the parameters from our discussion. I have the following thoughts:

1. The tolerances refer to different applications (admissibility of the solution, termination criteria, etc.), is it ok if 'eps' references all these tolerances? It may also make sense to set MIO tolerances separately (eps_mio?), as it makes no sense to set the parameters for non-mip.
2. Does a unit test for the eps feature make sense? If so, I would like to know what this looks like.

phschiele commented:
@Paulnkk You added the parameters to the `invert` function, which is called after the problem is solved. Instead, these tolerances should be applied before the solver is called.

Paulnkk commented:
@phschiele, thanks a lot for your support here! I merged changes into the branch. Should we use an existing problem to test for the eps parameter instead of the new one? I can also write a test that a specific parameter defined by the user is not overwritten by eps if this could increase coverage. Are there any additional changes I can make ?

phschiele commented:
@Paulnkk A few things before we can merge:
1. Adding the new keyword to the web docs
2. Make it work with enum parameters
3. Add the test you mentioned

Since we are testing a few different cases, let's keep this as a separate test case.

Paulnkk commented:
I need to check why this particular test is failing, not sure why since I just added the documentation. 

Also concerning the test I proposed, I am currently working on it but I am not sure how to access the parameters during and after the optimization since we do not return them to the user (pls correct me if I am wrong from my current understanding). I can make it work that the `solver_opts` are returned but that would make changes to productive code (actually pretty big changes if I think about it) and I am not sure if this is desired with the purpose of this PR.

phschiele commented:
@Paulnkk I would not make larger changes to the output in this PR. A simple (indirect) check is saving the error message from the test I've added, then including this parameter explicitly and checking that the error message changes.

Paulnkk commented:
@phschiele thanks for adjustments, was two minutes late to push haha, this was the test I wrote:


        first_error_message = ''
        first_error_param = ''

        try:
            prob.solve(solver=cp.MOSEK,
                       eps=1e-1,
                       mosek_params={'MSK_DPAR_INTPNT_CO_TOL_DFEAS': 1e-6})
        except mosek.Error as e:
            # Save the error message and mosek parameter
            first_error_message = str(e)
            first_error_param = str(e).split("for parameter '")[1].split("'")[0]
        try:
            prob.solve(solver=cp.MOSEK,
                       eps=1e-1,
                       mosek_params={'MSK_DPAR_INTPNT_CO_TOL_DFEAS': 1e-6,
                                     first_error_param: 1e-1})
        except mosek.Error as e:
            # Check that the error message is the same as before
            assert str(e) == first_error_message

phschiele commented:
@Paulnkk @rileyjmurray @SteveDiamond I pushed a few small changes to make this PR ready to merge from my side.

Since this is a mosek PR from a fork, we should get in the habit of verifying all tests passing before merging, which I did for this PR.

Verified that the docs render as well:
![image](https://github.com/cvxpy/cvxpy/assets/44360364/b13325b2-ce88-43c0-b031-481d59b2cf3f)

phschiele commented:
> @phschiele thanks for adjustments, was two minutes late to push haha, this was the test I wrote (please let me know if I understood it correctly):

Yes, that was about what I had in mind. I think the new test I've added achieves the same coverage and was a bit shorter to write.
I removed the enum you introduced mainly to avoid confusion with the existing mosek enums, and changed from a set to a tuple to have the same ordering each time.

SteveDiamond commented on cvxpy/reductions/solvers/conic_solvers/mosek_conif.py:
```diff
@@ -472,7 +475,7 @@ def invert(self, solver_output, inverse_data):
         solver_opts = solver_output['solver_options']
 
-        if solver_opts.get('accept_unknown', False):
+        if solver_opts['accept_unknown']:
```
Is accept_unknown always present?

SteveDiamond reviewed (approved):
LGTM
